Serve multistream graph captures from correct pool

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114647

This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

zdevito committed:
Serve multistream graph captures from correct pool

This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

[ghstack-poisoned]

zdevito committed:
Update on "Serve multistream graph captures from correct pool"


This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zdevito committed:
Update on "Serve multistream graph captures from correct pool"


This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zdevito committed:
Update on "Serve multistream graph captures from correct pool"


This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zdevito committed:
Update on "Serve multistream graph captures from correct pool"


This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ngimel reviewed (approved):
Thanks!

ngimel commented:
would it make sense to add a test?

eellison commented:
doing review this morning sorry for delay

eellison commented on aten/src/ATen/cuda/CUDAGraph.cpp:
```diff
@@ -136,7 +136,12 @@ void CUDAGraph::capture_begin(MempoolId_t pool/*=0*/, cudaStreamCaptureMode capt
   // due to the capture status being updated _after_ a capture had already started.
-  c10::cuda::CUDACachingAllocator::beginAllocateStreamToPool(capture_dev_, capture_stream_, mempool_id_);
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(capture_dev_, mempool_id_, [this](cudaStream_t stream) {
+      cudaStreamCaptureStatus status;
+      CaptureId_t stream_capture_id;
+      AT_CUDA_CHECK(cudaStreamGetCaptureInfo(stream, &status, &stream_capture_id));
+      return status == cudaStreamCaptureStatus::cudaStreamCaptureStatusActive && stream_capture_id == capture_id_;
```
So just to confirm, 

```
with torch.cuda.stream(s0):
  g.capture_begin()
  z = x+y
  with torch.cuda.stream(s1):
    s1.wait_stream(s0)
    w = z+y
  s0.wait_stream(s1)
```

For both the `z` and `w` allocation we will get the same `capture_id_` ?

ngimel replied:
Yes

eellison commented on aten/src/ATen/cuda/CUDAGraph.cpp:
```diff
@@ -136,7 +136,12 @@ void CUDAGraph::capture_begin(MempoolId_t pool/*=0*/, cudaStreamCaptureMode capt
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(capture_dev_, mempool_id_, [this](cudaStream_t stream) {
+      cudaStreamCaptureStatus status;
+      CaptureId_t stream_capture_id;
+      AT_CUDA_CHECK(cudaStreamGetCaptureInfo(stream, &status, &stream_capture_id));
```
We're calling `cudaStreamGetCaptureInfo` a lot more now, but that probably doesn't really matter, and this is off the hot path anyway

ngimel replied:
We call it only when there are captures underway, the way we used to

eellison reviewed (approved):
looks good - won't necessarily cover the non capturing case for when we are doing warmup but nonetheless we should land

ngimel commented:
Should this be landed?

zdevito commented:
I wanted to add a test that checks this behavior keeps working, but I haven't had time to write it yet.

zdevito committed:
Update on "Serve multistream graph captures from correct pool"


This fixes #114320 by placing the logic for determining whether to allocate
to a pool inside a callback that is controlled by CUDAGraph.cpp or by the
python bound api to allocate a stream directly to a pool.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

eellison commented:
land ?

zdevito commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "breaking multiple internal build jobs, please check internal diff in order to obtain more details" -c nosignal
